### PR TITLE
9792: Run inlineCallbacks callbacks outside of the _DefGen_Return exception context

### PR DESCRIPTION
**Pull request by mstojcevich**
## Scope and purpose

Python 3's implicit exception chaining will chain `_DefGen_Return` or `StopIteration` onto exceptions raised during `inlineCallbacks`-decorated function's callbacks because those exceptions are in the current exception context while the callbacks are running.

This PR changes `inlineCallbacks` to run the callbacks outside of the `_DefGen_Return`/`StopIteration` exception contexts.

## Contributor Checklist:

* [X] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/9792
* [X] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [X] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [X] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [X] I have updated the automated tests and checked that all checks for the PR are green.
* [X] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [X] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: mstojcevich, twm
Reviewer: graingert
Fixes: ticket:9792, ticket:10262

Run inlineCallbacks callbacks outside of the _DefGen_Return exception context

Python 3's implicit exception chaining will chain _DefGen_Return or StopIteration onto exceptions raised during inlineCallbacks-decorated function's callbacks because those exceptions are in the current exception context while the callbacks are running. This changes inlineCallbacks to run callbacks outside of the context of _DefGen_Return and StopIteration exceptions.
```

### CODE REVIEW

File: src/twisted/internet/defer.py [hunk 1]
```diff
@@ -1661,8 +1663,8 @@ def _inlineCallbacks(
                 result = current_context.run(gen.send, result)
         except StopIteration as e:
             # fell off the end, or "return" statement
-            status.deferred.callback(getattr(e, "value", None))
-            return
+            should_callback = True
+            callback_val = getattr(e, "value", None)
```
Comment (graingert) [inline src/twisted/internet/defer.py]: I think this is always present from 3.3+

```suggestion
            callback_val = e.value
```

Comment (graingert) [reply src/twisted/internet/defer.py]: Asyncio also sets the result in the except block https://github.com/python/cpython/blob/3.9/Lib/asyncio/tasks.py#L265

File: src/twisted/test/test_defer.py [hunk 2]
```diff
@@ -3557,3 +3558,96 @@ async def testFunction() -> bool:
...
+        there should not be any exception context if there wasn't any to begin
+        with.
+        """
+        for do_nothing in (self.doNothingReturn, self.doNothingReturnValue):
+            yield do_nothing()
+            # There were never any exceptions that this function was aware of
+            # (_DefGen_Return and StopIteration are implementation details
+            #  of 'do_nothing' and shouldn't leak into the exc_info of the
+            #  caller)
+            self.assertEqual(sys.exc_info(), (None, None, None))
```
Comment (graingert) [inline src/twisted/test/test_defer.py:3687]: Ah amazing I've always seen weird StopIteration related Tracebacks and had no idea why

**Review by graingert — CHANGES_REQUESTED** [review]
Can you test this with async/await also please

File: src/twisted/internet/defer.py [hunk 3]
```diff
@@ -1735,12 +1737,16 @@ def _inlineCallbacks(
                     lineno,
                 )
 
-            status.deferred.callback(e.value)
-            return
+            should_callback = True
```
Comment (graingert) [inline src/twisted/internet/defer.py]: I think this is cleaner with `break`


```python
while 1:
    ...
    callbackVal = e.value
    break
    ...
    callbackVal = e.value
    break

status.deferred.callback(callbackVal)
```

Comment (graingert) [reply src/twisted/internet/defer.py]: @mstojcevich here's the change https://github.com/twisted/twisted/commit/39ee52c3db975308a0d0ba336d4adb2673d48c5d

**Comment by graingert** [comment]
@mstojcevich these tests pass already without the changes to `defer.py`: https://github.com/twisted/twisted/pull/1654

**Comment by mstojcevich** [comment]
> @mstojcevich these tests pass already without the changes to `defer.py`: #1654

Oh, interesting. I originally wrote these having them sleep instead of yielding something that does nothing. I think having a real callback that came back from the reactor might have been important to have them fail without the changes. I'll see if I can get them working right.

**Comment by exarkun** [comment]
Just noting there is a lot of overlap between this PR and https://github.com/twisted/twisted/pull/1662

**Comment by glyph** [comment]
@mstojcevich I'm sorry we don't have a ton of review bandwidth these days, but would you mind resolving conflicts and addressing the feedback here?

**Comment by twm** [comment]
I have merged forward and added tests based on the reproducer attached to [Trac #9792](https://twistedmatrix.com/trac/ticket/9792).

**Comment by adiroiban** [comment]
@twm there is also #1662 , maybe we can have that merged first as it looks to me like it overlaps a bit

**Comment by graingert** [comment]
@twm can you demo that the tests fail without the fix by reverting the fix on a new branch?

**Comment by adiroiban** [comment]
I think that the demo part should be done by the reviewer.

As part of the review for #1662 I did just that, kept the tests, reverted the implementation changes, the test failed

**Review by adiroiban — CHANGES_REQUESTED** [review]
I am not sure if we still need this.

#1662 was merged and this PR now has conflicts.

Do we still need this PR?

thanks!

**Comment by twm** [comment]
@adiroiban The fix in #1662 was equivalent, so we don't need the changes here. There are additional tests here, but as they didn't actually reproduce the bug I'm not sure of their value. At the least conflicts would need to be resolved, and the 0.1 second sleep in the tests reduced to 0.0. I don't plan to do any of that myself &mdash; my goal was to get whichever PR across the finish line.

**Comment by twm** [comment]
Given the above I will close this for now, but if you are interested in landing these tests please feel free to reopen @mstojcevich!

**Pull request closed without merging**